src/UsgsAstroPushFrameSensorModel.cpp

```diff
@@ -0,0 +1,2461 @@
+
+  // Parameter covariance, hardcoded accuracy values
+  // hardcoded ~1 pixel accuracy values
+  int num_params = NUM_PARAMETERS;
+  double positionVariance = m_gsd * m_gsd;
+  // parameter time is scaled to [0, 2]
+  // so divide by 2 for velocities and 4 for accelerations
+  double velocityVariance = positionVariance / 2.0;
+  double accelerationVariance = positionVariance / 4.0;
+  m_covariance.assign(num_params * num_params, 0.0);
+
+  // Set position variances
+  m_covariance[0] = positionVariance;
+  m_covariance[num_params + 1] = positionVariance;
+  m_covariance[2 * num_params + 2] = positionVariance;
+  m_covariance[3 * num_params + 3] = velocityVariance;
+  m_covariance[4 * num_params + 4] = velocityVariance;
+  m_covariance[5 * num_params + 5] = velocityVariance;
+
+  // Set orientation variances
+  m_covariance[6 * num_params + 6] = positionVariance;
+  m_covariance[7 * num_params + 7] = positionVariance;
+  m_covariance[8 * num_params + 8] = positionVariance;
+  m_covariance[9 * num_params + 9] = velocityVariance;
+  m_covariance[10 * num_params + 10] = velocityVariance;
+  m_covariance[11 * num_params + 11] = velocityVariance;
+  m_covariance[12 * num_params + 12] = accelerationVariance;
+  m_covariance[13 * num_params + 13] = accelerationVariance;
+  m_covariance[14 * num_params + 14] = accelerationVariance;
+
+  // Set focal length variance
+  m_covariance[15 * num_params + 15] = positionVariance;
+}
+
+//---------------------------------------------------------------------------
+// Core Photogrammetry
+//---------------------------------------------------------------------------
+
+//***************************************************************************
+// UsgsAstroPushFrameSensorModel::groundToImage
+//***************************************************************************
+csm::ImageCoord UsgsAstroPushFrameSensorModel::groundToImage(
+    const csm::EcefCoord& ground_pt, double desired_precision,
+    double* achieved_precision, csm::WarningList* warnings) const {
+  MESSAGE_LOG(
+      "Computing groundToImage(No adjustments) for {}, {}, {}, with desired "
+      "precision {}",
+      ground_pt.x, ground_pt.y, ground_pt.z, desired_precision);
+
+  // The public interface invokes the private interface with no adjustments.
+  return groundToImage(ground_pt, _no_adjustment, desired_precision,
+                       achieved_precision, warnings);
+}
+
+//***************************************************************************
+// UsgsAstroPushFrameSensorModel::groundToImage (internal version)
+//***************************************************************************
+csm::ImageCoord UsgsAstroPushFrameSensorModel::groundToImage(
+    const csm::EcefCoord& groundPt, const std::vector<double>& adj,
+    double desiredPrecision, double* achievedPrecision,
+    csm::WarningList* warnings) const {
+
+  // Binary search for the framelet that is closest to the given ground point.
```

Comment:
We can probably do better than a binary search, but the search space is small enough (at most 100,000 framelets) that it converges fast enough.